mm2src/mm2_event_stream/src/manager.rs

```diff
@@ -0,0 +1,451 @@
+                .get_mut(streamer_id)
+                .ok_or(StreamingManagerError::StreamerNotFound)?
+                .clients
+                .remove(&client_id);
+
+            // If there are no more listening clients, terminate the streamer.
+            if streamers.get(streamer_id).map(|info| info.clients.len()) == Some(0) {
+                streamers.remove(streamer_id);
+            }
+            Ok(())
+        } else {
+            Err(StreamingManagerError::UnknownClient)
+        }
+    }
+
+    /// Broadcasts some event to clients listening to it.
+    ///
+    /// In contrast to `StreamingManager::send`, which sends some data to a streamer,
+    /// this method broadcasts an event to the listening *clients* directly, independently
+    /// of any streamer (i.e. bypassing any streamer).
+    pub fn broadcast(&self, event: Event) {
+        if let Some(client_ids) = self.streamers.read().get(event.origin()).map(|info| &info.clients) {
+            self.controller.read().broadcast(event, Some(client_ids))
+        }
+    }
+
+    /// Forcefully broadcasts an event to all known clients even if they are not listening for such an event.
+    pub fn broadcast_all(&self, event: Event) { self.controller.read().broadcast(event, None) }
+
+    pub fn new_client(&self, client_id: u64) -> Result<Receiver<Arc<Event>>, StreamingManagerError> {
+        let mut clients = self.clients.lock().unwrap();
+        if clients.contains_key(&client_id) {
+            return Err(StreamingManagerError::ClientExists);
+        }
+        clients.insert(client_id, HashSet::new());
+        // Note that events queued in the channel are `Arc<` shared.
+        // So a 1024 long buffer isn't actually heavy on memory.
+        Ok(self.controller.write().create_channel(client_id, 1024))
+    }
+
+    pub fn remove_client(&self, client_id: u64) -> Result<(), StreamingManagerError> {
+        let mut clients = self.clients.lock().unwrap();
+        // Remove the client from our known-clients map.
+        let listening_to = clients.remove(&client_id).ok_or(StreamingManagerError::UnknownClient)?;
+        // Remove the client from all the streamers it was listening to.
+        let mut streamers = self.streamers.write();
+        for streamer_id in listening_to {
+            if let Some(streamer_info) = streamers.get_mut(&streamer_id) {
+                streamer_info.clients.remove(&client_id);
+            } else {
+                log::error!(
+                    "Client {client_id} was listening to a non-existent streamer {streamer_id}. This is a bug!"
+                );
+            }
+            // If there are no more listening clients, terminate the streamer.
+            if streamers.get(&streamer_id).map(|info| info.clients.len()) == Some(0) {
+                streamers.remove(&streamer_id);
+            }
+        }
+        // Remove our channel with this client.
+        self.controller.write().remove_channel(&client_id);
+        Ok(())
+    }
```

Comment:
refer to the commit msg here: https://github.com/KomodoPlatform/komodo-defi-framework/pull/2172/commits/94621f3b23880ac4380ce982f8267fdb292175c0

so basically, to ensure atomicity at any point, we want to keep the locks for both mutexes (clients & streamers) held for the whole period till all updates are performed. so if we did the updates on the clients map we don't drop the mutex yet until updates are also done on the streamers map.

The solution for such a case is to bundle them in one struct and have them wrapped with only one mutex since they have a high data dependency. but i favored leaving each of them in a separate field because you can then easily get a read access over the streamers which is supposed to be the most used operation here (internal comm i.e. `send` needs read access, while external comm i.e. `API:add/remove-clients/streamers` needs write access).

also, when locking multiple mutexes from different threads, an order needs to be specified to avoid deadlocks, so here always clients is locked first then streamers.

p.s. the controller internal state also have a dependency with the clients map (every id in the controller has a corresponding id in the clients map), which after writing this comment btw i now think this is very loosely coupled and we might be better off merging the controller inside the clients map espcially since it's now less complicated (async channel drop functionality was removed).

Reply:
im now thinking of wrapping the whole state in a rwlock since having clients only in a mutex alone doesn't really give extra perf or something. will check that when i get my hands on the code.

Reply:
ok cool and thank you for the detailed answer. I think grouping them will be good for manageability too.